### ISSUE TITLE
Extracting elements on List of Maps throws exception on unknown fields

### ISSUE THREAD

**Issue by filiphr** [assertj/assertj-core#2420]
#### Summary
We recently upgraded AssertJ to 3.21.0, we skipped 3.20 (because we got the upgrade through Spring Boot from 2.5 to 2.6) and we started getting failures in places that were working before.
This is caused by the fix that was done in #2111. 

The example below runs fine with 3.19. However, it fails on 3.21.

In my honest opinion the fact that  `hasFieldOrProperty` used to do `map.get(name)` is correct. If I am extracting values from a map, I might be extracting sparse values. The fact that other frameworks are using a Map as the actual object and not overriding the method properly is entirely up to the frameworks to fix that on their side. 

#### Example

```java
List<Map<String, Object>> elements = new ArrayList<>();
Map<String, Object> map = new HashMap<>();
map.put("name", "Yoda");
elements.add(map);
map = new HashMap<>();
map.put("name", "Luke");
map.put("location", "Tatooine");

assertThat(elements)
        .extracting("name", "location")
        .containsExactlyInAnyOrder(tuple("Yoda", null), tuple("Luke", "Tatooine"));

```

There is a workaround for this if I use `.extracting(m -> m.get("name"), m -> m.get("location"))`. However, that is really verbose and I would argue that it is against the spirit of AssertJ

**Comment by filiphr** [comment]
Forgot to add something #2111 was originally about `hasFieldOrProperty` where I guess it is fine to throw an exception when the map does not contain the key. However, when extracting I would not expect the failure.

**Comment by scordio** [comment]
The main actor for all these cases is `getSimpleValue`:
https://github.com/assertj/assertj-core/blob/a9a93f0858db73eb15560011687fb12a6975bcba/src/main/java/org/assertj/core/util/introspection/PropertyOrFieldSupport.java#L65-L94

We've already realized that this method is involved in too many flows and we're doing too much here, so we're planning some refactoring to make sure each assertion is flexible enough.

However, I'm not sure if the refactoring can fit under 3.22.0 as we already have quite a long list of changes for that version. Would it be ok for you to temporarily go with the workaround, but have it properly addressed latest in 3.23.0?

**Comment by filiphr** [comment]
>We've already realized that this method is involved in too many flows and we're doing too much here, so we're planning some refactoring to make sure each assertion is flexible enough.

Yes I remember seeing some issues around this. I even think I chimed in some of them 😄.

>However, I'm not sure if the refactoring can fit under 3.22.0 as we already have quite a long list of changes for that version. Would it be ok for you to temporarily go with the workaround, but have it properly addressed latest in 3.23.0?

We rolled back to 3.19.0 for now. I'll see if I can do something like a PR in this area. Do you by any chance remember which issue you had about the refactoring?

**Comment by scordio** [comment]
We briefly discussed this topic with @joel-costigliola, but we didn't create a corresponding issue yet.

For example, I had in mind to introduce an implicit conversion for `extracting` methods having a `String` parameter, so something like this would work:

```java
assertThat(instance).extracting("field1.0.field3").isEqualTo(...) // field1 is a Map<Integer, Object>
```

However, this could cause side effects to other features, like the recursive comparison.

My initial thought about refactoring would be to start splitting `getSimpleValue` into strategies, and maybe `getValueOf` as well. Then we should evaluate the next steps, looking at the usage patterns. Ideally, we should have the flexibility to slightly modify the behavior of each assertion/feature without affecting the others.

My time is quite limited in the next days but I should be able to dig more next week. Of course, any ideas are very much appreciated 🙂

**Comment by joel-costigliola** [comment]
@filiphr we are thinking to let users to choose the behavior when extracting an unknown key from a map.

The question is what should be the default behavior, returning null as in 3.19.0 or failing as in the current version?
I don't have a strong opinion yet, my concern with returning null is that you can't differentiate unknown keys from the ones that exist with a null value.
Thoughts?

**Comment by scordio** [comment]
I vote for keeping the failure as the default behavior, mostly because of https://github.com/assertj/assertj-core/issues/2111#issuecomment-770436485.

Also, I'm not sure if we really want to have a global flag that would affect different assertion patterns.

I'll look into it during the weekend and see if maybe we can decouple the recursive comparison from the other assertions without big changes.

**Comment by filiphr** [comment]
In my opinion `extracting` should just do `Map#get` and not try to check if the map contains the key, i.e. the same as it was in 3.19.0. 

I am not sure that it is worth it to have a special behaviour for choosing how to extract an unknown key from a map. I think that the `extracting` and `hasFieldOrProperty` decoupling as mentioned by @scordio is the way to go.

**Comment by joel-costigliola** [comment]
that could be an interesting option to explore, thanks for the suggestion @pilak
